compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp

```diff
@@ -113,6 +113,163 @@ struct FoldAttentionMaskUnitDim final
+    // `collapse_shape` op.
+    llvm::SmallDenseSet<int64_t> toDrop;
+    for (const auto &[outDim, indices] : llvm::enumerate(collapseReassoc)) {
+      for (auto [innerIdx, inDim] : llvm::enumerate(indices)) {
+        // Can't drop this dim if it isnt statically 1 or if it isn't being
+        // combined with any other dimensions.
+        if (indices.size() == 1 || interShape[inDim] != 1) {
+          continue;
+        }
+
+        // If we are collapsing multiple unit dims together, at least 1 must be
+        // kept (prefer the first).
+        if (outShape[outDim] == 1 && innerIdx != 0) {
+          continue;
+        }
+        toDrop.insert(inDim);
+      }
+    }
+
+    // Step 2: remove dimensions from `toDrop` that weren't introduced by the
+    // `expand_shape` op.
+    const auto expandReassoc = expandOp.getReassociationIndices();
+    for (const auto &[inDim, indices] : llvm::enumerate(expandReassoc)) {
+      if (indices.size() == 1) {
+        toDrop.erase(indices[0]);
+      }
+    }
+    if (toDrop.empty()) {
+      return rewriter.notifyMatchFailure(collapseOp,
+                                         "Didn't find any unit dims to drop");
+    }
+
+    // Step 3: construct a new intermediate shape without the foldable dims.
+    SmallVector<int64_t> newInterShape;
+    SmallVector<OpFoldResult> newOutputSizes;
+    newInterShape.reserve(interShape.size() - toDrop.size());
+    newOutputSizes.reserve(interShape.size() - toDrop.size());
+    SmallVector<OpFoldResult> outputSizes = getMixedValues(
+        expandOp.getStaticOutputShape(), expandOp.getOutputShape(), rewriter);
+    for (auto [idx, ofr] : llvm::enumerate(outputSizes)) {
+      if (!toDrop.contains(idx)) {
+        std::optional<int64_t> staticDim = getConstantIntValue(ofr);
+        newInterShape.push_back(staticDim.value_or(ShapedType::kDynamic));
+        newOutputSizes.push_back(ofr);
+      }
+    }
+
+    /// Attempts to push back a new `ReassociationIndices` to `reassoc` that
+    /// does not include the dimensions in `toDrop`.
+    /// Returns true if new `ReassociationIndices` were appended to `reassoc`.
+    auto pushBackReassociation =
+        [&toDrop](SmallVectorImpl<ReassociationIndices> &reassoc, int64_t start,
+                  int64_t count, int64_t origStart) {
+          ReassociationIndices indices;
+          indices.reserve(count);
+          int64_t dim = start;
+          for (int64_t idx : llvm::seq<int64_t>(origStart, origStart + count)) {
+            if (!toDrop.contains(idx)) {
+              indices.push_back(dim++);
+            }
+          }
+          if (!indices.empty()) {
+            reassoc.push_back(std::move(indices));
+            return true;
+          }
+          return false;
+        };
+
+    auto isIdentityReassociation = [](ArrayRef<ReassociationIndices> reassoc) {
+      return llvm::all_of(reassoc,
+                          [](auto &indices) { return indices.size() == 1; });
```

Comment:
Can this be `llvm::all_of(reassoc, llvm::hasSingleElement)`? btw, maybe move it right before where it is used, i.e., step 6. This way, they can be in the same screen when people browse the code. I think it is one of intentions that makes it lambda, but not a static method?